plugins/catalog-backend/src/service/createRouter.ts

```diff
@@ -135,24 +137,139 @@ export async function createRouter(
+        let writeLock: DeferredPromise | undefined;
+        const controller = new AbortController();
+        req.on('end', () => {
+          controller.abort();
+          writeLock?.resolve();
+          writeLock = undefined;
+        });
+        res.on('drain', () => {
+          writeLock?.resolve();
+          writeLock = undefined;
+        });
+
+        res.setHeader('Content-Type', 'application/json; charset=utf-8');
+        res.status(200);
+        res.flushHeaders();
+        res.write(prettyPrint ? '[\n' : '[');
+
+        const limit = 10000;
+        let cursor: Cursor | undefined;
+        let firstSend = true;
+        do {
+          const result = await entitiesCatalog.queryEntities(
+            !cursor
+              ? { credentials, fields, limit, filter, orderFields: order }
+              : { credentials, fields, limit, cursor },
+          );
+
+          if (result.items.length) {
+            // TODO(freben): This is added as a compatibility guarantee, until we can be
+            // sure that all adopters have re-stitched their entities so that the new
+            // targetRef field is present on them, and that they have stopped consuming
+            // the now-removed old field
+            // TODO(jhaals): Remove this in April 2022
+            for (const entity of result.items) {
+              if (entity.relations) {
+                for (const relation of entity.relations as any) {
+                  if (!relation.targetRef && relation.target) {
+                    // This is the case where an old-form entity, not yet stitched with
+                    // the updated code, was in the database
+                    relation.targetRef = stringifyEntityRef(relation.target);
+                  } else if (!relation.target && relation.targetRef) {
+                    // This is the case where a new-form entity, stitched with the
+                    // updated code, was in the database but we still want to produce
+                    // the old data shape as well for compatibility reasons
+                    relation.target = parseEntityRef(relation.targetRef);
+                  }
+                }
+              }
+            }
+
+            // Stringify
+            let data: string;
+            if (prettyPrint) {
+              data = JSON.stringify(result.items, null, 2).slice(2, -2);
+              if (!firstSend) {
+                data = `,\n${data}`;
+              }
+            } else {
+              data = JSON.stringify(result.items).slice(1, -1);
+              if (!firstSend) {
+                data = `,${data}`;
+              }
+            }
+
+            if (writeLock) {
+              await writeLock;
+            }
+
+            if (controller.signal.aborted) {
+              res.end();
+              return;
+            }
+
+            if (!res.write(data)) {
+              // The kernel buffer is full. Create the lock but do not await it
```

Comment:
Will `data` be lost in this case? 🤔

Reply:
No, this is just an optimization. We just stream out blocks as fast as the reader consumes them, instead of as fast as we can produce them (which fills up our memory when the consumer is slow). This should actually deal with tricky scenarios where large numbers of consumers simultaneously consume the entire database, even if they are relatively slow.

Reply:
nice nice